security/tpm/src/tpm.ts

```diff
@@ -3,22 +3,428 @@
+                  if (!ekCreateErr) {
+                    self._ek = ekPublicKey;
+                    self._createPersistentPrimary(new PersKeyInfo('SRK', tss.Owner, srkPersHandle, srkTemplate), (srkCreateErr: Error, srkPublicKey: TPMT_PUBLIC) => {
+                      if (!srkCreateErr) {
+                        self._srk = srkPublicKey;
+                        self._fsm.transition('connected', callback);
+                      } else {
+                        self._fsm.transition('disconnected', callback, srkCreateErr);
+                      }
+                    });
+                  } else {
+                    self._fsm.transition('disconnected', callback, ekCreateErr);
+                  }
+                });
+              });
+            } catch (err) {
+              this._fsm.transition('disconnected', callback, err);
+            }
+          },
+          '*': () => this._fsm.deferUntilTransition()
+        },
+        connected: {
+          _onEnter: (callback, err) => {
+            callback(null);
+          },
+          getEndorsementKey: (callback) => {
+            callback(null, this._ek.asTpm2B());
+          },
+          getStorageRootKey: (callback) => {
+            callback(null, this._srk.asTpm2B());
+          },
+          signWithIdentity: (deviceIdData, callback) => {
+            this._signData(deviceIdData, (errSign: Error, signedData: Buffer) => {
+              if (errSign) {
+                debug('Error from signing data: ' + errSign);
+                callback(errSign);
+                this._fsm.transition('disconnected');
+              } else {
+                callback(null, signedData);
+              }
+            });
+          },
+          activateSymmetricIdentity: (identityKey, callback) => {
+            this._activateSymetricIdentity(identityKey, (errActivate: Error) => {
+              if (errActivate) {
+                debug('Error from activate: ' + errActivate);
+                callback(errActivate);
+                this._fsm.transition('disconnected');
+              } else {
+                callback(null, null);
+              }
+            });
+          },
+        }
+      }
+    });
+  }
+
+  getEndorsementKey(callback: (err: Error, endorsementKey: string) => void): void {
+      this._fsm.handle('getEndorsementKey', callback);
+  }
+
+  getStorageRootKey(callback: (err: Error, storageKey: string) => void): void {
+    this._fsm.handle('getStorageRootKey', callback);
+  }
+
+  signWithIdentity(deviceIdData: Buffer, callback: (err: Error, signedResponse: Buffer) => void): void {
+    if (deviceIdData === null || deviceIdData.length === 0) {
+        throw new ReferenceError('\'deviceIdData\' cannot be \'' + deviceIdData + '\'');
+    }
+
+    if (this._idKeyPub == null) {
+        throw new errors.ArgumentError('activateIdentityKey first before signing');
```

Comment:
>ArgumentError [](start = 25, length = 13)

this._idKeyPub is not an argument so this is probably not the right error. InvalidOperationError maybe? also the error would benefit from being a proper sentence.

Reply:
Done.

---
In reply to: [157889459](https://github.com/Azure/azure-iot-sdk-node/pull/185#discussion_r157889459) [](ancestors = 157889459)